framework/src/main/java/org/checkerframework/common/wholeprograminference/SceneToStubWriter.java

```diff
@@ -0,0 +1,439 @@
+
+    /** How far to indent when writing members of a stub file. */
+    private static final String INDENT = "    ";
+
+    /** The printer where the stub file is written. */
+    private final PrintWriter printWriter;
+
+    /**
+     * A map from the descriptions of ATypeElement_s to the corresponding base Java types, since
+     * AScene_s don't carry that information. See the comment on the {@code basetypes} field of
+     * {@link WholeProgramInferenceScenesHelper} for more information.
+     */
+    private Map<String, TypeMirror> basetypes;
+
+    /**
+     * A map from fully-qualified class names to the TypeElement that represents them. Computed by
+     * {@link WholeProgramInferenceScenes}. Used to output the correct names of generic parameters
+     * to classes, which the stub format requires.
+     */
+    private Map<String, TypeElement> types;
+
+    /**
+     * The fully-qualified name of enums to be written to a stub file.
+     *
+     * <p>The stub parser can't parse an enum that's labeled "class", but {@link AClass} doesn't
+     * specify if a class is an enum. So track which classes are enums.
+     */
+    private Set<String> enumSet;
+
+    /**
+     * Stolen from {@code IndexFileWriter}. Prints a literal; used when printing the arguments of
+     * annotations.
+     */
+    private void printValue(AnnotationFieldType aft, Object o) {
+        if (aft instanceof AnnotationAFT) {
+            printAnnotation((Annotation) o);
+        } else if (aft instanceof ArrayAFT) {
+            ArrayAFT aaft = (ArrayAFT) aft;
+            printWriter.print('{');
+            if (!(o instanceof List)) {
+                printValue(aaft.elementType, o);
+            } else {
+                List<?> l = (List<?>) o;
+                // watch out--could be an empty array of unknown type
+                // (see AnnotationBuilder#addEmptyArrayField)
+                if (aaft.elementType == null) {
+                    if (l.size() != 0) {
+                        throw new AssertionError();
+                    }
+                } else {
+                    boolean first = true;
+                    for (Object o2 : l) {
+                        if (!first) {
+                            printWriter.print(',');
+                        }
+                        printValue(aaft.elementType, o2);
+                        first = false;
+                    }
+                }
+            }
+            printWriter.print('}');
+        } else if (aft instanceof ClassTokenAFT) {
+            printWriter.print(aft.format(o));
+        } else if (aft instanceof BasicAFT && o instanceof String) {
+            printWriter.print(Strings.escape((String) o));
+        } else if (aft instanceof BasicAFT && o instanceof Long) {
+            printWriter.print(o.toString() + "L");
+        } else {
+            printWriter.print(o.toString());
+        }
+    }
+
+    /** Prints an annotation in Java source format. */
```

Comment:
Does Annotation#toString() not print it in proper format?

Reply:
It gets the long formatting I mentioned above wrong. I don't know if it's correct, otherwise.